src/pybind/mgr/balancer/module.py

```diff
@@ -1008,7 +1008,7 @@ def do_crush_compat(self, plan):
 
         # Make sure roots don't overlap their devices.  If so, we
         # can't proceed.
-        roots = pe.target_by_root.keys()
+        roots = list(pe.target_by_root.keys())
```

Comment:
According to [1] this is not thread-safe (?). That article suggests the following as an alternative - maybe it could be considered if this code is not performance-sensitive?

```
roots = list(pe.target_by_root.copy().keys())
```